src/Client/Request.php

```diff
@@ -138,7 +146,16 @@ public function handleData($data)
         // buffer until double CRLF (or double LF for compatibility with legacy servers)
         if (false !== strpos($this->buffer, "\r\n\r\n") || false !== strpos($this->buffer, "\n\n")) {
             try {
-                list($response, $bodyChunk) = $this->parseResponse($this->buffer);
+                $psrResponse = gPsr\parse_response($this->buffer);
+
+                if ($this->responseIsAnUpgradeResponse($psrResponse)) {
+                    $this->stream->removeListener('data', array($this, 'handleData'));
+
+                    $this->emit('upgrade', array($this->stream, $psrResponse, $this));
```

Comment:
Not sure about the whole `upgrade` event. How does this interfere with the existing request/response semantics?

I like how this is exposed on the server side via https://github.com/reactphp/http#streaming-outgoing-response, perhaps we can find a similar solution here?

Reply:
I actually followed a Node piece I read [here](https://nodejs.org/api/http.html#http_event_upgrade) as I also mentioned above.

I will check out the sample you sent, I kind of like it, knowing that a `ThroughStream` also implements` DuplexStreamInterface`.